File: src/main/java/no/nav/veilarbperson/utils/PersonVisittkortDataMapper.java
package no.nav.veilarbperson.utils;

import no.nav.veilarbperson.client.pdl.HentPerson;
import no.nav.veilarbperson.client.pdl.domain.Diskresjonskode;
import no.nav.veilarbperson.client.pdl.domain.Telefon;
import no.nav.veilarbperson.domain.PersonVisittkortData;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import static no.nav.veilarbperson.utils.PersonV2DataMapper.*;

public class PersonVisittkortDataMapper {

    private PersonVisittkortDataMapper() {}

    public static PersonVisittkortData tilPersonVisittkortData(
            HentPerson.Person person,
            boolean erSkjermet,
            String krrTelefon,
            String krrTelefonOppdatert) {

        Optional<HentPerson.Navn> navn = hentGjeldeneNavn(person.getNavn());

        List<Telefon> telefoner = new ArrayList<>(mapTelefonNrFraPdl(person.getTelefonnummer()));
        leggKrrTelefonIListe(krrTelefon, krrTelefonOppdatert, telefoner);

        return new PersonVisittkortData(
                navn.map(HentPerson.Navn::getFornavn).orElse(null),
                navn.map(HentPerson.Navn::getMellomnavn).orElse(null),
                navn.map(HentPerson.Navn::getEtternavn).orElse(null),
                Optional.ofNullable(getFirstElement(person.getFoedselsdato()))
                        .map(HentPerson.Foedselsdato::getFoedselsdato)
                        .map(Object::toString)
                        .orElse(null),
                Optional.ofNullable(getFirstElement(person.getDoedsfall()))
                        .map(HentPerson.Doedsfall::getDoedsdato)
                        .map(Object::toString)
                        .orElse(null),
                Optional.ofNullable(getFirstElement(person.getKjoenn()))
                        .map(HentPerson.Kjoenn::getKjoenn)
                        .orElse(null),
                Optional.ofNullable(getFirstElement(person.getAdressebeskyttelse()))
                        .map(HentPerson.Adressebeskyttelse::getGradering)
                        .map(Diskresjonskode::mapKodeTilTall)
                        .orElse(null),
                erSkjermet,
                Optional.ofNullable(getFirstElement(person.getSikkerhetstiltak()))
                        .map(HentPerson.Sikkerhetstiltak::getBeskrivelse)
                        .orElse(null),
                telefoner
        );
    }

    /** KRR-telefon får alltid prioritet 1. PDL-nummer med samme nummer fjernes. Øvrige PDL-prioriteter bumpes. */
    static void leggKrrTelefonIListe(String krrTelefon, String oppdatert, List<Telefon> liste) {
        if (krrTelefon == null) return;
        liste.removeIf(t -> krrTelefon.equals(t.getTelefonNr()));
        liste.stream()
                .filter(t -> !"KRR".equals(t.getMaster()))
                .forEach(t -> t.setPrioritet(String.valueOf(Integer.parseInt(t.getPrioritet()) + 1)));
        liste.add(0, new Telefon().setPrioritet("1").setTelefonNr(krrTelefon).setRegistrertDato(oppdatert).setMaster("KRR"));
    }
}

File: src/main/java/no/nav/veilarbperson/utils/PersonVisittkortDataMapper.kt
package no.nav.veilarbperson.utils

import no.nav.veilarbperson.client.pdl.HentPerson
import no.nav.veilarbperson.client.pdl.domain.Diskresjonskode
import no.nav.veilarbperson.client.pdl.domain.Telefon
import no.nav.veilarbperson.domain.PersonVisittkortData

object PersonVisittkortDataMapper {

    fun tilPersonVisittkortData(
        person: HentPerson.Person,
        erSkjermet: Boolean,
        krrTelefon: String?,
        krrTelefonOppdatert: String?
    ): PersonVisittkortData {
        val navn = PersonDataMapper.hentGjeldeneNavn(person.navn).orElse(null)

        val telefoner = PersonDataMapper.mapTelefonNrFraPdl(person.telefonnummer).toMutableList()
        leggKrrTelefonIListe(krrTelefon, krrTelefonOppdatert, telefoner)

        return PersonVisittkortData(
            fornavn = navn?.fornavn,
            mellomnavn = navn?.mellomnavn,
            etternavn = navn?.etternavn,
            fodselsdato = PersonDataMapper.getFirstElement(person.foedselsdato)?.foedselsdato?.toString(),
            dodsdato = PersonDataMapper.getFirstElement(person.doedsfall)?.doedsdato?.toString(),
            kjonn = PersonDataMapper.getFirstElement(person.kjoenn)?.kjoenn,
            diskresjonskode = PersonDataMapper.getFirstElement(person.adressebeskyttelse)
                ?.gradering
                ?.let { Diskresjonskode.mapKodeTilTall(it) },
            egenAnsatt = erSkjermet,
            sikkerhetstiltak = PersonDataMapper.getFirstElement(person.sikkerhetstiltak)?.beskrivelse,
            telefon = telefoner
        )
    }

    /** KRR-telefon får alltid prioritet 1. PDL-nummer med samme nummer fjernes. Øvrige PDL-prioriteter bumpes. */
    private fun leggKrrTelefonIListe(krrTelefon: String?, oppdatert: String?, liste: MutableList<Telefon>) {
        if (krrTelefon == null) return
        liste.removeIf { krrTelefon == it.telefonNr }
        liste.filter { it.master != "KRR" }
            .forEach { it.setPrioritet((it.prioritet.toInt() + 1).toString()) }
        liste.add(0, Telefon().setPrioritet("1").setTelefonNr(krrTelefon).setRegistrertDato(oppdatert).setMaster("KRR"))
    }
}
